Experimental command to compare resource to its other API version

mikhailshilkov opened the pull request:
Resolves https://github.com/pulumi/pulumi-azure-native/issues/2197

We want to use tools like this to calculate which versions of Azure Native we can deprecate. Sample commands:

Compare two versions of the same resource, see if there are forward-incompatible changes (i.e., code upgrade may not compile automatically):
```
./schema-tools squeeze --old azure-native:resources/v20180201:ResourceGroup --new azure-native:resources/v20220901:ResourceGroup
Found 1 breaking change:
Resource "azure-native:resources/v20220901:ResourceGroup" missing input "name"
```

Go through all versions of a resource, try upgrading each version to the next one, remove all versions that are forward-compatible:
```
./schema-tools squeeze -r resources:ResourceGroup
All versions:
azure-native:resources/v20151101:ResourceGroup
azure-native:resources/v20160201:ResourceGroup
azure-native:resources/v20160701:ResourceGroup
azure-native:resources/v20160901:ResourceGroup
azure-native:resources/v20170510:ResourceGroup
azure-native:resources/v20180201:ResourceGroup
azure-native:resources/v20180501:ResourceGroup
azure-native:resources/v20190301:ResourceGroup
azure-native:resources/v20190501:ResourceGroup
azure-native:resources/v20190510:ResourceGroup
azure-native:resources/v20190701:ResourceGroup
azure-native:resources/v20190801:ResourceGroup
azure-native:resources/v20191001:ResourceGroup
azure-native:resources/v20200601:ResourceGroup
azure-native:resources/v20200801:ResourceGroup
azure-native:resources/v20201001:ResourceGroup
azure-native:resources/v20210101:ResourceGroup
azure-native:resources/v20210401:ResourceGroup
azure-native:resources/v20220901:ResourceGroup
Not forward-compatible versions:
azure-native:resources/v20180201:ResourceGroup
azure-native:resources/v20220901:ResourceGroup
```

Do the same for all resources, print summaries:
```
./schema-tools squeeze                                                                                                          
aad:DomainService can be reduced from 8 to 3
aad:OuContainer can be reduced from 7 to 1
aadiam:DiagnosticSetting can be reduced from 3 to 1
aadiam:PrivateEndpointConnection can be reduced from 2 to 1
aadiam:azureADMetric can be reduced from 2 to 1
aadiam:privateLinkForAzureAd can be reduced from 3 to 1
addons:SupportPlanType can be reduced from 3 to 1
advisor:Suppression can be reduced from 6 to 2
agfoodplatform:Extension can be reduced from 3 to 2
agfoodplatform:FarmBeatsModel can be reduced from 3 to 2
agfoodplatform:PrivateEndpointConnection can be reduced from 1 to 1
agfoodplatform:Solution can be reduced from 1 to 1
alertsmanagement:ActionRuleByName can be reduced from 3 to 2
alertsmanagement:AlertProcessingRuleByName can be reduced from 2 to 1
alertsmanagement:PrometheusRuleGroup can be reduced from 1 to 1
alertsmanagement:SmartDetectorAlertRule can be reduced from 4 to 1
analysisservices:ServerDetails can be reduced from 5 to 1
apimanagement:Api can be reduced from 14 to 4
apimanagement:ApiDiagnostic can be reduced from 14 to 3
apimanagement:ApiDiagnosticLogger can be reduced from 3 to 2
apimanagement:ApiIssue can be reduced from 14 to 1
apimanagement:ApiIssueAttachment can be reduced from 14 to 1
apimanagement:ApiIssueComment can be reduced from 14 to 1
apimanagement:ApiManagementService can be reduced from 16 to 4
...
```

It also writes a file `replacements.json` where keys are versions being deprecated and values being the versions that they are compatible to and will remain in the version set.

danielrbradley commented:
Recap of code walkthrough:
- The main output we're looking for is a list of resources which are candidates for deprecation and removal because they can be safely migrated.
- Should we consider changes to which properties are required as breaking changes?
  - Inputs matter when they become required.
  - Outputs matter when then become optional.
  - What's the impact to the potential deprecations if we exclude these cases?
  - Given we're doing this during a major version change, perhaps it's okay for users to have to include new properties?

mikhailshilkov commented:
I updated this PR to print a flat list of all compatible resources. Also, I added a check for required properties changes, and they only eliminated about 1% of extra resources. Here is a dump of the command output, it’s at 9870 rows: https://drive.google.com/file/d/1RGO5WsnI1p4yXOzLI0lTUhud0usob7o1/view?usp=sharing

danielrbradley commented:
Wondering if we could include in the output the version that each deprecated resource should be migrated to - so we can populate deprecation notices with instructions

mikhailshilkov commented:
@danielrbradley We probably can. Right now the tool compares with the next version, which can then be eliminated too, so we'd have to be smart and find the last version that remains after all eliminations, but that should be doable.

What's your preferred format - I assume we want to keep the file machine-readable?

danielrbradley commented:
JSON output is probably the most portable as we'll be feeding the output into other tools. It should also make it easy to add additional properties in the future without breaking existing usages.

danielrbradley reviewed (commented):
This looks good so far. It would be ideal to include for each version being deprecated - the remaining version it can be upgraded to. Though this isn't required as we could do this during the version calculation process in azure-native if needed.

danielrbradley reviewed (approved):
Think this is good to ship then we can test and can always do follow-up work where required
